pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,628 @@
+		partial = len(v1Rules)
+	case !compat && len(v1Rules) > 0 && len(v2Rules) > 0: // Unexpected newer rules found.
+		partial = len(v2Rules)
+	}
+
+	// Exit if there's nothing to do. Return the count of partially migrated
+	// rules.
+	switch {
+	case !compat && len(v1Rules) == 0 && len(v2Rules) > 0: // 2
+		fallthrough
+	case compat && len(v1Rules) > 0 && len(v2Rules) == 0: // 3
+		return 0, 0, partial
+	}
+
+	isUpgrade := !compat && len(v1Rules) > 0  // 1
+	isDowngrade := compat && len(v2Rules) > 0 // 4
+
+	// The following operation will be done on a per-rule basis (or
+	// per-endpoint, assuming that each egress rule has a unique IP addr
+	// associated with the endpoint).
+	//
+	// In both the upgrade and downgrade scenario, the following happens in a
+	// specific order to guarantee that any failure at any point won't cause
+	// connectivity disruption for the endpoint. Any errors encountered do not
+	// stop the migration process because we want to ensure that we conform to
+	// either the new state or the old state, and want to avoid being
+	// in-between datapath states.
+	//   1) Copy over new routes from the old routes
+	//   2) Insert new rule
+	//   3) Delete old rule
+	//   4) Delete old routes
+	// Doing (1) & (2) before (3) & (4) allows us to essentially perform an
+	// "atomic" swap-in for the new state.
+	//
+	// If a failure occurs at (1), then the old state can continue to service
+	// the endpoint. Similarly with (2) because routes without rules are likely
+	// to not have any effect.
+	//
+	// If a failure occurs, (3) or (4), then we already have the new state
+	// in-place to direct traffic for the endpoint. There is a slight nuance
+	// here; the older rule has a _higher_ priority which means it would be
+	// selected by Linux first, before reaching the new rule. Therefore, if we
+	// fail at (3), keep in mind we do have the new state in-place, but it goes
+	// "untouched". We haven't deleted the routes and, therefore the old state
+	// is still in-place as well, servicing the endpoint and allowing no
+	// connectivity disruptions. This is the main reason why the new priority
+	// was chosen to have a lower priority than the original priority value.
+	//
+	// If we fail at (4), then the old rule will have
+	// been deleted and the new state is in-place, which would be servicing the
+	// endpoint. The old routes would just be leftover state to be cleaned up
+	// at a later point.
+	//
+	// It is also important to note that we only revert what we've done on a
+	// per-rule basis if we fail at (2) or (3). This is by design because we
+	// want to ensure that each iteration of the loop is atomic to each
+	// endpoint. Meaning, either the endpoint ends up with the new datapath or
+	// the original.
+
+	// Number of rules (endpoints) successfully migrated and how many failed.
+	var migrated, failed int
+
+	if isUpgrade {
+		for _, r := range v1Rules {
```

Comment:
Now that there's less state sharing between iterations of this loop, is it possible to factor the inner logic into a separate `m.upgradeRule()` function so that we can reason about this function from a high level of interactions between iterations of the loop, then deal with a single-endpoint case in a dedicated function?

Reply:
Note that if the function return an error, then the `failed++` / `continue` statements would not need to be duplicated either, the top level loop would look something like:

```
        if isUpgrade {
                for _, r := range v1Rules {
                        if err := m.upgradeRule(r); err != nil {
                                log.WithField("rule", r).Warning("...")
                                failed++
                                continue
                        }
                        migrated++
                }
        } else if isDowngrade {
...
```